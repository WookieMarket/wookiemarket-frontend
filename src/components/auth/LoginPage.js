import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../store/selectors";
import { resetError } from "../../store/slices/ui";
import { authLogin } from "../../store/slices/auth";
import ErrorModal from "../shared/modal/ErrorModal";
import Spinner from "../shared/spinner/Spinner";
import Layout from "../layout/Layout";

//DONE Log in with username and password and a checkbox to give the option to persist the token, also handle errors and user feedback. When doing Login I want to send the user to the page they wanted to go to.

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    username: "",
    password: "",
    rememberMe: false,
  });

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    setCredentials(prevCredentials => ({
      ...prevCredentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <Layout title="Login Page">
      <div>
        {isLoading ? (
          <Spinner message="cargando..." />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <label className="form-label" htmlFor="username">
              Username
            </label>
            <input
              id="username"
              className="form-input"
              type="text"
              name="username"
              onChange={handleChange}
              value={credentials.username}
              required
            />
            <label className="form-label" htmlFor="password">
              Password
            </label>
            <input
              id="password"
              className="form-input"
              type="password"
              name="password"
              onChange={handleChange}
              value={credentials.password}
              required
            />
            <label htmlFor="rememberMe">
              rememberMe
              <input
                id="rememberMe"
                type="checkbox"
                name="rememberMe"
                checked={credentials.rememberMe}
                onChange={handleChange}
              />
            </label>
            <button
              data-testid="button"
              type="submit"
              //variant="primary"
              width="button-form"
              disabled={buttonDisabled}>
              Log in
            </button>
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={handleErrorClick}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
