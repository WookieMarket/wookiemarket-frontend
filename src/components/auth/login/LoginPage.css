.container-form {
  width: 50%;
  height: 50vh; /* Establece la altura del contenedor en el 50% de la altura de la ventana */
  position: absolute; /* Cambia a posici√≥n absoluta */
  top: 55%; /* Alinea el borde superior del contenedor al 50% de la ventana */
  left: 50%; /* Alinea el borde izquierdo del contenedor al 50% de la ventana */
  transform: translate(-50%, -50%); /* Centra vertical y horizontalmente */
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 30px 10px 20px 10px;
  gap: 10px;
  border-radius: 10px;
  text-align: center;
  border: 2px solid var(--border-color);
  background-color: var(--background-color-form);
}

.container-form .form-group {
  width: 100%;
  padding: 0 15px;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.container-form .password-label {
  font-family: var(--font-family);
  font-size: 13px;
  margin-bottom: 10px;
}

.container-form .password-input {
  height: 20px;
  width: 100%;
  font-family: var(--font-family);
  margin: auto;
  font-size: 12px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}
.container-form .email-input {
  height: 20px;
  width: 100%;
  margin: auto;
  font-size: 13px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}

.container-form .link-reset {
  display: flex;
  justify-content: center;
  font-family: var(--font-family);
  font-size: 14px;
  color: rgba(0, 0, 255, 0.897);
  text-decoration: none;
  margin-bottom: 10px;
}

.container-form .password-input-rememberme {
  display: flex;
  font-family: var(--font-family);
  align-items: center;
  align-content: space-between;
  font-size: 14px;
}

@media screen and (min-width: 768px) {
  .container-form {
    width: 40%;
  }
  .container-form .password-input {
    height: 25px;
    font-size: 12px;
  }
  .container-form .password-input-rememberme {
    font-size: 14px;
  }
  /* .container-form .login-button {
    border-radius: 5px;
    padding: 7px 7px;
    font-size: 15px;
  } */
  .container-form .link-reset {
    font-size: 13px;
  }
  .container-form .password-label {
    font-size: 14px;
    margin-bottom: 10px;
  }
}

@media screen and (min-width: 1024px) {
  .container-form {
    width: 30%;
  }
  .container-form .password-input {
    height: 30px;
    font-size: 14px;
  }
  .container-form .link {
    font-size: 14px;
  }

  .container-form .password-input-rememberme {
    font-size: 14px;
  }
  /* .container-form .login-button {
    border-radius: 5px;
    padding: 8px 8px;
    font-size: 15px;
  } */
  .container-form .password-label {
    font-size: 16px;
    margin-bottom: 10px;
  }
}
