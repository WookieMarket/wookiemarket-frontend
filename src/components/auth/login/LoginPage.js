import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../../store/selectors";
import { resetError, toggleModal } from "../../../store/slices/ui";
import { authLogin, emailResetPassword } from "../../../store/slices/auth";
import ErrorModal from "../../shared/modal/ErrorModal";
import Spinner from "../../shared/spinner/Spinner";
import Layout from "../../layout/Layout";
import { useTranslation } from "react-i18next";
import Modal from "../../shared/modal/Modal";
import { Link } from "react-router-dom";
import Form from "../../shared/form/Form";
import "./LoginPage.css";
import Button from "../../shared/Button";

//DONE Log in with username and password and a checkbox to give the option to persist the token, also handle errors and user feedback. When doing Login I want to send the user to the page they wanted to go to.

function LoginPage() {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const { isLoading, error, showModal } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    username: "",
    password: "",
    rememberMe: false,
  });

  const [email, setEmail] = useState("");

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    setCredentials(prevCredentials => ({
      ...prevCredentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));
  };

  const handleResetPasswordLinkClick = () => {
    dispatch(toggleModal());
  };

  const handleShowModalconfirm = async event => {
    event.preventDefault();
    dispatch(emailResetPassword(email));
    dispatch(toggleModal());
  };

  const handleShowModalCancel = () => {
    dispatch(toggleModal());
  };

  const handleChangeEmail = e => {
    setEmail(e.target.value);
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <Layout title={t("Login Page")}>
      <div>
        {isLoading ? (
          <Spinner message={t("charging...")} />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <Form
              classNameForm="form-group"
              classNameLabel="password-label"
              htmlFor="username"
              text={t("Username")}
              classNameInput="password-input"
              inputId="username"
              inputType="text"
              inputName="username"
              value={credentials.username}
              handleChange={handleChange}
              placeholder={t("Username")}
              required
            />

            <Form
              classNameForm="form-group"
              classNameLabel="password-label"
              htmlFor="password"
              text={t("Password")}
              classNameInput="password-input"
              inputId="password"
              inputType="password"
              inputName="password"
              value={credentials.password}
              handleChange={handleChange}
              placeholder={t("Password")}
              required
            />

            <Form
              classNameForm="password-input-rememberme"
              htmlFor="rememberMe"
              text={t("RememberMe")}
              inputId="rememberMe"
              inputType="checkbox"
              inputName="rememberMe"
              checked={credentials.rememberMe}
              handleChange={handleChange}
              required
            />

            <Button
              //className="login-button"
              data-testid="button"
              type="submit"
              variant="accept"
              width="button-form"
              disabled={buttonDisabled}>
              {t("Log in")}
            </Button>
            <Link onClick={handleResetPasswordLinkClick}>
              <h4 className="link-reset">{t("Forgot password?")}</h4>
            </Link>
            {showModal && (
              <Modal
                title={t("Recover password")}
                message={
                  <Form
                    classNameForm="form-group"
                    classNameLabel="password-label"
                    htmlFor="email-service"
                    text={t("Enter your email")}
                    classNameInput="email-input"
                    inputId="email-service"
                    inputType="email"
                    inputName="reset-email"
                    value={email}
                    handleChange={handleChangeEmail}
                    placeholder={t("Email")}
                    required
                  />
                }
                onConfirm={handleShowModalconfirm}
                onCancel={handleShowModalCancel}
              />
            )}
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={handleErrorClick}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
