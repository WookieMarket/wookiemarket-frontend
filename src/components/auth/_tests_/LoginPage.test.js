import { render, screen } from "@testing-library/react";
import LoginPage from "../LoginPage";

import { configureStore } from "@reduxjs/toolkit";

import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import authReducer, { authLogin } from "../../../store/slices/auth";
import uiReducer, { resetError } from "../../../store/slices/ui";

jest.mock("../../../store/slices/auth");

console.log("authred", authReducer);
console.log("login", authLogin);

describe("LoginPage", () => {
  const renderComponent = (error = null) => {
    const store = configureStore({
      reducer: {
        auth: authReducer,
        ui: uiReducer, // Agrega el slice uiReducer al store
      },
      preloadedState: {
        auth: false,
        ui: { error }, // Establece el estado de error en el slice ui
      },
    });
    return render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>,
    );
  };

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("shoul dispatch authLogin action", () => {
    const credentials = {
      username: "rober",
      password: "123",
      rememberMe: true,
    };

    //NOTE renderizo el componente
    renderComponent();

    const usernameInput = screen.getByLabelText(/Username/);
    const passwordInput = screen.getByLabelText(/Password/);
    const checkboxInput = screen.getByLabelText(/rememberMe/);
    const submitButton = screen.getByRole("button", { name: /Log in/ });

    expect(submitButton).toBeDisabled();

    //NOTE para lanzar eventos

    userEvent.type(usernameInput, credentials.username);
    userEvent.type(passwordInput, credentials.password);
    userEvent.click(checkboxInput, credentials.rememberMe);

    expect(submitButton).toBeEnabled();

    userEvent.click(submitButton);
    console.log(authLogin);
    expect(authLogin).toHaveBeenCalledWith(credentials);
  });

  test("should display an error", () => {
    const error = { message: "Unauthorized" };
    renderComponent(error);

    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();

    const modalButton = screen.getByTestId("modalButton");

    userEvent.click(modalButton);

    expect(resetError).toHaveBeenCalled();
  });
});
