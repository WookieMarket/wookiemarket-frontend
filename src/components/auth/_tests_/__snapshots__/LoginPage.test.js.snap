// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div>
    <header
      class="header"
    >
      <nav
        class="navbar"
      >
        <h1
          class="title"
        >
           
          header
           
        </h1>
        <div
          class="icons"
        >
          <img
            alt="ES"
            class="icon-language"
            src="/images/languageIcons/es.png"
            title="icon-hover"
          />
          <img
            alt="EN"
            class="icon-language"
            src="/images/languageIcons/en.png"
            title="icon-hover"
          />
        </div>
        <button
          as="[object Object]"
          to="/login"
        >
          Login
        </button>
      </nav>
      <hr />
    </header>
    <main
      class="content"
    >
      <h2
        class="layout-title"
      >
        Login Page
      </h2>
      <div>
        <form
          class="container-form"
        >
          <label
            class="form-label"
            for="username"
          >
            Username
          </label>
          <input
            class="form-input"
            id="username"
            name="username"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-input"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <label
            for="rememberMe"
          >
            RememberMe
            <input
              id="rememberMe"
              name="rememberMe"
              type="checkbox"
            />
          </label>
          <button
            data-testid="button"
            disabled=""
            type="submit"
            width="button-form"
          >
            Log in
          </button>
        </form>
      </div>
    </main>
    <footer>
      <hr />
      <div
        class="footer"
      >
        <p>
          footer
        </p>
      </div>
    </footer>
  </div>
</div>
`;
